File: AdmissionServerApp/app/src/main/java/com/example/adeeliftikhar/admissionserverapp/MainActivity.java
package com.example.adeeliftikhar.admissionserverapp;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.example.adeeliftikhar.admissionserverapp.Internet.CheckInternetConnectivity;
import com.example.adeeliftikhar.admissionserverapp.SessionManager.LoginSessionManager;
import com.google.firebase.auth.FirebaseAuth;

public class MainActivity extends AppCompatActivity {
    DrawerLayout drawerLayout;
    ActionBarDrawerToggle toggleButton;
    LoginSessionManager loginSessionManager;
    NavigationView navigationView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        toolbarSet();
        drawerLayoutSet();
        navigationViewWork();
    }

    //        End of onCreate Method.
    private void toolbarSet() {
//        Custom Toolbar used as ActionBar...
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
    }

    private void drawerLayoutSet() {
//        Navigation Drawer...
        drawerLayout = findViewById(R.id.drawer_layout);
        toggleButton = new ActionBarDrawerToggle(MainActivity.this, drawerLayout, R.string.text_open, R.string.text_close);
        drawerLayout.addDrawerListener(toggleButton);
        toggleButton.syncState();
    }

    private void navigationViewWork() {
        navigationView = findViewById(R.id.navigation_view);
        navigationView.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
                int id = menuItem.getItemId();
                switch (id) {
                    case R.id.chairman_message:
                        if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                        } else {
                            Intent intentMessage = new Intent(MainActivity.this, FragmentsContainerActivity.class);
                            String message = "message";
                            intentMessage.putExtra("id", message);
                            startActivity(intentMessage);
                        }
                        break;
                    case R.id.team:
                        if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                        } else {
                            Intent intentTeam = new Intent(MainActivity.this, FragmentsContainerActivity.class);
                            String team = "team";
                            intentTeam.putExtra("id", team);
                            startActivity(intentTeam);
                        }
                        break;
                    case R.id.history:
                        if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                        } else {
                            Intent intentHistory = new Intent(MainActivity.this, FragmentsContainerActivity.class);
                            String history = "history";
                            intentHistory.putExtra("id", history);
                            startActivity(intentHistory);
                        }
                        break;
                    case R.id.schedule:
                        if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                        } else {
                            Intent intentSchedule = new Intent(MainActivity.this, FragmentsContainerActivity.class);
                            String schedule = "schedule";
                            intentSchedule.putExtra("id", schedule);
                            startActivity(intentSchedule);
                        }
                        break;
                    case R.id.rules:
                        if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                        } else {
                            Intent intentRules = new Intent(MainActivity.this, FragmentsContainerActivity.class);
                            String rules = "rules";
                            intentRules.putExtra("id", rules);
                            startActivity(intentRules);
                        }
                        break;
                    case R.id.facility:
                        if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                        } else {
                            Intent intentFacility = new Intent(MainActivity.this, FragmentsContainerActivity.class);
                            String facility = "facility";
                            intentFacility.putExtra("id", facility);
                            startActivity(intentFacility);
                        }
                        break;
                    case R.id.activities:
                        if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                        } else {
                            Intent intentActivities = new Intent(MainActivity.this, FragmentsContainerActivity.class);
                            String activities = "activities";
                            intentActivities.putExtra("id", activities);
                            startActivity(intentActivities);
                        }
                        break;
                    case R.id.scholarship_info:
                        if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                        } else {
                            Intent intentInformation = new Intent(MainActivity.this, FragmentsContainerActivity.class);
                            String information = "information";
                            intentInformation.putExtra("id", information);
                            startActivity(intentInformation);
                        }
                        break;
                    case R.id.merit_list:
                        if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                        } else {
                            Intent intentList = new Intent(MainActivity.this, FragmentsContainerActivity.class);
                            String list = "list";
                            intentList.putExtra("id", list);
                            startActivity(intentList);
                        }
                        break;
                }
                return true;
            }
        });
    }

    public void seeAdmissions(View view) {
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.logout_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (toggleButton.onOptionsItemSelected(item)) {
            return true;
        }
        if (toggleButton.onOptionsItemSelected(item)) {
            return true;
        }
        if (id == R.id.logout) {
            logoutAlertBox();
        }
        return super.onOptionsItemSelected(item);
    }

    private void logoutAlertBox() {
        AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
        builder.setTitle("Logout");
        builder.setMessage("Are you sure to Logout?");
        builder.setPositiveButton("Logout", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                if (!CheckInternetConnectivity.isConnected(MainActivity.this)) {
                    Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
                } else {
                    logout();
                }
            }
        }).setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
            }
        });
        builder.show();
    }

    private void logout() {
        LoginSessionManager loginSessionManager = new LoginSessionManager(MainActivity.this);
        loginSessionManager.loginTheUser(false, "", "");
        FirebaseAuth.getInstance().signOut();
        Intent intent = new Intent(MainActivity.this, AdminLoginActivity.class);
        startActivity(intent);
        finish();
    }
}

File: Messenger/app/src/main/java/com/example/adeeliftikhar/messenger/MainActivity.kt
package com.example.adeeliftikhar.messenger

import android.app.ProgressDialog
import android.content.Intent
import android.graphics.drawable.Drawable
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.annotation.DrawableRes
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.app.AlertDialog
import android.support.v7.widget.DividerItemDecoration
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import com.example.adeeliftikhar.messenger.Internet.CheckInternetConnectivity
import com.example.adeeliftikhar.messenger.Models.ChatMessageModel
import com.example.adeeliftikhar.messenger.Models.RegisterModel
import com.example.adeeliftikhar.messenger.SessionsManager.LoginSessionManager
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.squareup.picasso.Picasso
import com.xwray.groupie.GroupAdapter
import com.xwray.groupie.Item
import com.xwray.groupie.ViewHolder
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.latest_message_row.view.*

class MainActivity : AppCompatActivity() {
    private val adapter = GroupAdapter<ViewHolder>()
    private val latestMessageMap = HashMap<String, ChatMessageModel>()

    companion object {
        var currentUser: RegisterModel? = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        spin_kit_view_latest_message.visibility = View.VISIBLE
        recycler_view_latest_messages.adapter = adapter
        recycler_view_latest_messages.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))

        adapter.setOnItemClickListener { item, view ->
            val intent = Intent(this, ChatLogActivity::class.java)
            val row = item as LatestMessageRow
            intent.putExtra(NewMessageActivity.USER_KEY, row.chatPartnerUser)
            startActivity(intent)
        }

        getLatestMessage()
        fetchCurrentUser()
    }

    private fun refreshRecyclerViewMessage() {
        adapter.clear()
        latestMessageMap.values.forEach {
            adapter.add(LatestMessageRow(chatMessage = it))
        }
        if (adapter.itemCount != 0) {
            spin_kit_view_latest_message.visibility = View.GONE
        }
    }

    private fun getLatestMessage() {
        val fromID = FirebaseAuth.getInstance().uid
        val dbRef = FirebaseDatabase.getInstance().getReference("/LatestMessage/$fromID")
        dbRef.keepSynced(true)
        dbRef.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (!snapshot.exists()) {
                    text_view_no_chat.visibility = View.VISIBLE
                    spin_kit_view_latest_message.visibility = View.GONE
                } else {
                    dbRef.addChildEventListener(object : ChildEventListener {
                        override fun onCancelled(p0: DatabaseError) {
                        }

                        override fun onChildMoved(p0: DataSnapshot, p1: String?) {
                        }

                        override fun onChildChanged(p0: DataSnapshot, p1: String?) {
                            val chatMessage = p0.getValue(ChatMessageModel::class.java) ?: return
                            latestMessageMap[p0.key!!] = chatMessage
                            refreshRecyclerViewMessage()
                        }

                        override fun onChildAdded(p0: DataSnapshot, p1: String?) {
                            val chatMessage = p0.getValue(ChatMessageModel::class.java) ?: return
                            latestMessageMap[p0.key!!] = chatMessage

                            refreshRecyclerViewMessage()
                        }

                        override fun onChildRemoved(p0: DataSnapshot) {
                        }
                    })
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })
    }

    private fun fetchCurrentUser() {
        val uid = FirebaseAuth.getInstance().uid
        val dbRef = FirebaseDatabase.getInstance().getReference("/UsersInfo/$uid")
        dbRef.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onCancelled(p0: DatabaseError) {
            }

            override fun onDataChange(p0: DataSnapshot) {
                currentUser = p0.getValue(RegisterModel::class.java)
            }

        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.action_bar_nav_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.new_message -> {
                if (!CheckInternetConnectivity.isConnected(this)) {
                    Toast.makeText(this, "No Internet Connection", Toast.LENGTH_SHORT).show()
                } else {
                    val intent = Intent(this, NewMessageActivity::class.java)
                    startActivity(intent)
                }
            }
            R.id.profile -> {
                val intent = Intent(this, ProfileActivity::class.java)
                startActivity(intent)
            }
            R.id.logout -> {
                logoutAlertBox()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun logoutAlertBox() {
        val builder = AlertDialog.Builder(this@MainActivity)
        builder.setTitle("Logout")
        builder.setMessage("Do you want to Logout?")
        builder.setPositiveButton("Logout") { dialog, which ->
            if (!CheckInternetConnectivity.isConnected(this@MainActivity)) {
                Toast.makeText(this@MainActivity, "No Internet Connection", Toast.LENGTH_SHORT).show()
            } else {
                dialogBoxLogout()
                logout()
            }
        }.setNegativeButton("Cancel") { dialog, which -> }
        builder.show()
    }

    private fun dialogBoxLogout() {
        val progressDialog = ProgressDialog(this)
        progressDialog.setMessage("Logout in Progress.")
        progressDialog.setCancelable(false)
        progressDialog.show()
    }

    private fun logout() {
        val loginSessionManager = LoginSessionManager(this@MainActivity)
        loginSessionManager.loginTheUser(false, "", "")
        FirebaseAuth.getInstance().signOut()
        val intent = Intent(this@MainActivity, LoginActivity::class.java)
        startActivity(intent)
        finish()
    }
}
